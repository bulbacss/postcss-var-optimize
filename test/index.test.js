const fs = require('fs')
const { resolve } = require('path')
const postcss = require('postcss')

let plugin = require('./../index')

// eslint-disable-next-line no-extend-native
RegExp.prototype.toJSON = RegExp.prototype.toString;

async function run (file, options = {}, useSnapshot = false) {
  let input = fs.readFileSync(resolve(__dirname, 'input/' + file + '.css'))
  let result = await postcss([plugin(options)])
    .process(input, { from: undefined })
  if (useSnapshot) {
    expect(result.css).toMatchSnapshot(file + ' ' + JSON.stringify(options))
    expect(result.warnings()).toHaveLength(0)
  } else if (fs.existsSync(resolve(__dirname, 'output/' + file + '.css'))) {
    let output = fs.readFileSync(resolve(__dirname, 'output/' + file + '.css'))
    expect(result.css).toEqual(output.toString())
    expect(result.warnings()).toHaveLength(0)
  } else {
    fs.writeFileSync(resolve(__dirname, 'output/' + file + '.css'), result.css)
  }
}

describe('internal workings', () => {
  it('finds variables declared and modified', async () => {
    let opts = {test: {}}
    let css = 'a{ } :root{--a: 1;--b: 2} .theme{--b: 3}'

    await postcss([plugin(opts)]).process(css, {from: undefined})

    expect(Object.keys(opts.test.declared)).toHaveLength(2)
    expect(opts.test.declared.a).toEqual('1')
    expect(opts.test.declared.b).toBeTruthy()
    expect(Object.keys(opts.test.modified)).toHaveLength(1)
    expect(opts.test.modified.b).toBe('2')
  })
});

describe('simple test cases', () => {
  it('removes the unmodified variables and replace them', async () => {
    await run('simple-theme')
  })

  it('works with nested variables', async () => {
    await run('theme-with-nested-variables')
  })
});

describe('options', () => {
  it('uses whitelist', async () => {
    await run('simple-theme', {
      whitelist: ['a', /^b$/]
    }, true)
    await run('theme-with-nested-variables', {
      whitelist: /^b|c$/
    }, true)
  })

  it('uses blacklist', async () => {
    await run('simple-theme', {
      blacklist: ['a', /^b$/]
    }, true)
    await run('theme-with-nested-variables', {
      blacklist: /^b|c$/
    }, true)
  })
})

describe('advanced test cases', () => {
  it ('removes everything in a file without customisation', async () => {
    await run('bulma-no-theme')
  })
  it('works with complex frameworks', async () => {
    await run('bulma-with-theme')
  })
});
