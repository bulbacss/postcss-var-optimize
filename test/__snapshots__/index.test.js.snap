// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options uses blacklist: simple-theme {"blacklist":["a","/^b$/"]} 1`] = `
"a {
  border: var(--a) var(--b) 1px var(--a);
  background: var(--q);
  padding: 0
}

:root {
  --a: 1;
  --b: 2;
}

.theme {
  --b: 3;
}
"
`;

exports[`options uses blacklist: theme-with-nested-variables {"blacklist":"/^b|c$/"} 1`] = `
"a {
  border: var(--c) var(--b) 1%;
  background: var(--c);
  padding: 1%;
}

:root {
  --c: hsl(20deg, var(--b), 1%);
  --b: 3%;
}

.theme {
  --b: 3;
}
"
`;

exports[`options uses whitelist: simple-theme {"whitelist":["a","/^b$/"]} 1`] = `
"a {
  border: 1 var(--b) 1px 1;
  background: var(--q);
  padding: 0
}

:root {
  --b: 2;
  --c: 3;
}

.theme {
  --b: 3;
  --c: 3;
}
"
`;

exports[`options uses whitelist: theme-with-nested-variables {"whitelist":"/^b|c$/"} 1`] = `
"a {
  border: var(--d);
  background: hsl(20deg, var(--b), var(--a));
  padding: var(--a);
}

:root {
  --a: 1%;
  --d: hsl(20deg, var(--b), var(--a)) var(--b) var(--a);
  --b: calc(var(--a) + 2%);
}

.theme {
  --b: 3;
}
"
`;
